### PR TITLE
Highlevel Bindings tutorial

### PR DESCRIPTION
**Pull request by kodebach**
## Basics

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [x] Longer descriptions should be in documentation or in design decisions.
- [x] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

- [ ] I added unit tests
- [ ] I ran all tests locally and everything went fine
- [x] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] meta data is updated (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))

## Review

- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)
- [ ] Documentation is concise and good to read
- [ ] Examples are well chosen and understandable

### CODE REVIEW

File: doc/tutorials/highlevel-bindings.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    errors because of missing keys, since we have no way of knowing which keys should exist. <br/><br/>
+    There is an exception to this rule. Some target languages have a standard error concept, which not only forces the
+    user to handle arising errors, but also does so in a simple and concise way. An example of this is Rust. Its `Result`
+    type forces the user to handle errors and the `?` operator allows to do this in a concise way. <br/>
+    However, we still recommend to avoid errors as far as possible.
+- **Idomatic Code:** If the targt language has a concept of "idomatic code", the API should fall into that category.
+
+## When to write Bindings
+
+Based on the goals above, you should decided, whether it is possible to write a binding for the C high-level API while
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]:
```suggestion
Based on the goals above, you should decide, whether it is possible to write a binding for the C high-level API while
```

File: doc/tutorials/highlevel-bindings.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+preserving the goals.
+
+If you decide to write a binding, proceed with this tutorial.
+
+If not, designing an API to meet our goals is up to you. While designing the API you should keep in mind that you can
+always utilise the code-generator like the C API does.
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]: and also extend kdb gen as described below.

File: doc/tutorials/highlevel-bindings.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+Writing the binding for the `highlevel` library works the same way as writing a language binding for any part of
+Elektra. The only additional challenge is that the binding should still meet the same goals and (as far as possible)
+fulfill the same guarantees as the C API.
+
+Some languages like to split bindings into two parts. One version that maps the C API one-to-one and another part
+that builds on the first one. The second part then uses the languages additional features. Since our API isn't inteded
+to be used directly, but through generated code, it might not be necessary to write the second part. It may be
+sufficient to write (or generate) a one-to-one mirror of the C API and use that in the code-generator template.
+
+On the other hand, the generated code should still be understandable. So if writing a more idiomatic API on top of the
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]:
```suggestion
The generated code should still be understandable. So if writing a more idiomatic API on top of the
```

File: doc/tutorials/highlevel-bindings.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+direct mapping, significantly simplifies the generated code (and maybe also the template), you should write such an API.
+
+### Creating the Code-Generator Template
+
+How to create a template for `kdb gen` is detailed in [this tutorial](code-generator.md).
+
+The created template should support the same input keysets (and parent keys) as the `highlevel` template. If and how
+exactly you implement the advanced features (structs, unions, ...) is up to you. It might make sense for your language
+or it might not. Note: enums should always be supported (if your language has them) as they are one of the `type`
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]:
```suggestion
Note: enums should always be supported (if your language has them) as they are one of the `type`
```

File: doc/tutorials/highlevel-bindings.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+On the other hand, the generated code should still be understandable. So if writing a more idiomatic API on top of the
+direct mapping, significantly simplifies the generated code (and maybe also the template), you should write such an API.
+
+### Creating the Code-Generator Template
+
+How to create a template for `kdb gen` is detailed in [this tutorial](code-generator.md).
+
+The created template should support the same input keysets (and parent keys) as the `highlevel` template. If and how
+exactly you implement the advanced features (structs, unions, ...) is up to you. It might make sense for your language
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]:
```suggestion
exactly you implement the advanced features (structs, unions, ...) is up to you.
```

File: doc/tutorials/highlevel-bindings.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+or it might not. Note: enums should always be supported (if your language has them) as they are one of the `type`
+plugins types.
+
+For example in C++ the generated code could consist of nested structs with lots of overloaded operators. In that case
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]:
```suggestion
For example, in C++ the generated code could consist of nested structs with overloaded operators. In that case
```

**Comment by markus2330** [comment]
@raphi011 could you please read this tutorial. Do not hesitate to ask here if there are any questions.

**Comment by kodebach** [comment]
@markus IMO this is ready to merge, link-checker won't work (see https://github.com/ElektraInitiative/libelektra/pull/2915#issuecomment-526836525)

**Comment by markus2330** [comment]
@raphi011 can you please pick up the work now?

**Comment by raphi011** [comment]
> @raphi011 can you please pick up the work now?

Yes, I will look at it tonight.

**Comment by markus2330** [comment]
Thank you!

File: doc/tutorials/highlevel-bindings.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+preserving the goals.
+
+If you decide to write a binding, proceed with this tutorial.
+
+If not, designing an API to meet our goals is up to you. While designing the API you should keep in mind that you can
+always use code-generator (`kdb gen`) templates like the C API does.
+
+## How to write Bindings
+
+Since the C high-level API consits of a shared library API and a code-generated part, your binding will also have these
```
Comment (raphi011) [inline doc/tutorials/highlevel-bindings.md]: consists

**Comment by markus2330** [comment]
@raphi011: can you fix these typos in a PR together with your new sections about your experiences how to write a Go binding?

**Comment by raphi011** [comment]
Sure, I'll do it as soon as elektradv2 is merged :)

**Comment by markus2330** [comment]
Thank you!

> I'll do it as soon as elektradv2 is merged :)

As it has nothing to do with each other, you can easily work in parallel on it And it is only docu, so it will not distract you from the other task.

**Pull request merged**